serial/serialwin32.py

```diff
@@ -285,11 +285,20 @@ def read(self, size=1):
                     ctypes.byref(self._overlapped_read))
                 if not read_ok and win32.GetLastError() not in (win32.ERROR_SUCCESS, win32.ERROR_IO_PENDING):
                     raise SerialException("ReadFile failed ({!r})".format(ctypes.WinError()))
-                result_ok = win32.GetOverlappedResult(
-                    self._port_handle,
-                    ctypes.byref(self._overlapped_read),
-                    ctypes.byref(rc),
-                    True)
+                timeout = Timeout(self._timeout)
+                while True:
+                    # even with timeout infinite, return from kernel mode every 100ms to allow for things like KeyboardInterrupt to happen
+                    millis = win32.DWORD(int(min(timeout.time_left() or 0.1, 0.1) * 1000))
+                    result_ok = win32.GetOverlappedResultEx(
+                        self._port_handle,
+                        ctypes.byref(self._overlapped_read),
+                        ctypes.byref(rc),
+                        millis,
+                        False)
+                    if result_ok or win32.GetLastError() != win32.WAIT_TIMEOUT:
```

Comment:
non-zero result is success, in case of timeout we expect failure and a GetLastError value of WAIT_TIMEOUT